indico/migrations/versions/20230810_1300_0af8f63aa603_refactor_editing_revisions.py

```diff
@@ -0,0 +1,308 @@
+"""Refactor editing revisions
+
+Revision ID: 0af8f63aa603
+Revises: a59688f9ba40
+Create Date: 2023-08-07 09:44:11.504348
+"""
+
+import sqlalchemy as sa
+from alembic import context, op
+
+from indico.core.db.sqlalchemy import PyIntEnum
+from indico.core.db.sqlalchemy.custom.utcdatetime import UTCDateTime
+from indico.util.enum import IndicoIntEnum
+
+
+# revision identifiers, used by Alembic.
+revision = '0af8f63aa603'
+down_revision = 'a59688f9ba40'
+branch_labels = None
+depends_on = None
+
+
+class _RevisionType(IndicoIntEnum):
+    new = 1
+    ready_for_review = 2
+    needs_submitter_confirmation = 3
+    changes_acceptance = 4
+    changes_rejection = 5
+    needs_submitter_changes = 6
+    acceptance = 7
+    rejection = 8
+    replacement = 9
+    reset = 10
+
+
+class _FinalRevisionState(IndicoIntEnum):
+    none = 0
+    replaced = 1
+    needs_submitter_confirmation = 2
+    needs_submitter_changes = 3
+    accepted = 4
+    rejected = 5
+    undone = 6
+
+
+def _create_new_revisions(conn):
+    query = f'''
+        SELECT editable_id, submitter_id, editor_id, reviewed_dt, final_state, comment
+        FROM event_editing.revisions AS a
+        WHERE a.final_state IN ({_FinalRevisionState.needs_submitter_changes},
+                                {_FinalRevisionState.accepted},
+                                {_FinalRevisionState.rejected}) AND
+              NOT EXISTS (SELECT 1
+                          FROM event_editing.revisions AS b
+                          WHERE a.id != b.id AND
+                                a.editable_id = b.editable_id AND
+                                a.final_state = {_FinalRevisionState.needs_submitter_changes} AND
+                                b.final_state = {_FinalRevisionState.needs_submitter_changes} AND
+                                a.editor_id IS NOT NULL AND
+                                a.editor_id = b.editor_id AND
+                                a.reviewed_dt IS NOT NULL AND
+                                a.reviewed_dt = b.reviewed_dt)
+    '''
+    new_revisions = []
+    for editable_id, submitter_id, editor_id, reviewed_dt, final_state, comment in conn.execute(query):
+        revision_type = {
+            _FinalRevisionState.needs_submitter_changes: _RevisionType.changes_rejection,
```

Comment:
I believe this is what's causing the "submitter rejected proposed changes" bug. I think this should fix it:
```suggestion
            _FinalRevisionState.needs_submitter_changes: _RevisionType.needs_submitter_changes if editor_id else _RevisionType.changes_rejection,
```

Reply:
You probably want this since technically the ID can be `0`. This is incredibly unlikely since only legacy indico used 0 as a valid ID, but still ;)

```
            _FinalRevisionState.needs_submitter_changes: (_RevisionType.needs_submitter_changes
                                                          if editor_id is not None
                                                          else _RevisionType.changes_rejection),
```

Reply:
@duartegalvao will you commit the fix?

Reply:
sure, in just a sec